src/menelaus/concept_drift/md3.py

```diff
@@ -0,0 +1,358 @@
+        # The decision boundary is the line y = a*x - b
+        # Sources:
+        #    (1) MD3 Paper
+        #    (2) https://scikit-learn.org/stable/auto_examples/svm/plot_svm_margin.html#sphx-glr-download-auto-examples-svm-plot-svm-margin-py
+        #    (3) https://scikit-learn.org/stable/auto_examples/svm/plot_separating_hyperplane.html#example-svm-plot-separating-hyperplane-py
+        #    (4) https://stackoverflow.com/questions/23794277/extract-decision-boundary-with-scikit-learn-linear-svm
+        self.w = np.array(self.classifier.coef_[0])
+        self.intercept = np.array(self.classifier.intercept_)
+        self.a = -self.w[0] / self.w[1]
+        self.b = self.intercept[0] / self.w[1]
+
+        # calculate the magnitude of the margin
+        self.margin = 1 / np.sqrt(np.sum(self.classifier.coef_**2))
+
+    # TODO: two ways we can handle the update/drift detection:
+    #       (1) after the detector has been updated with some number of samples
+    #           that the user can pass in (maybe 30), then we start looking for drift
+    #       (2) the user sets the detector to have a reference batch, and then passes
+    #           in samples one at a time to update the detector
+    def set_reference(self, reference_batch, target_name):
+        """
+        Initialize detector with a reference batch. Reference batch must be
+        manually set and updated by user using this method. Reference batch
+        is not automatically updated after a drift is detected.
+
+        Args:
+            reference_batch (DataFrame): initial baseline dataset
+            target_name (string): name of the column in the reference_batch
+                dataframe which is the target variable
+        """
+
+        self.reference_batch_features = reference_batch.loc[:, reference_batch.columns != target_name]
+        self.reference_batch_target = reference_batch.loc[:, reference_batch.columns == target_name]
+        
+        self.reference_distribution = self.calculate_distribution_statistics(reference_batch)
+        
+        if self.oracle_data_length_required is None:
+            self.oracle_data_length_required = self.reference_distribution["len"]
+
+        self.forgetting_factor = (self.reference_distribution["len"] - 1) / self.reference_distribution["len"]
+        self.curr_margin_density = self.reference_distribution["md"]
+
+    def calculate_distribution_statistics(self, data):
+        """
+        Calculate the following five statistics for the data distribution
+        passed in:
+            1. Length/Number of Samples (len)
+            2. Margin Density (md)
+            3. Standard Deviation of Margin Density (md_std)
+            4. Accuracy (acc)
+            5. Standard Deivation of Accuracy (acc_std)
+
+        Args:
+            data (DataFrame): batch of data to calculate distribution
+                statistics for
+        """
+
+        duplicate_classifier = self.classifier
```

Comment:
I did not, but I agree using `sklearn.base.clone` would be safer. Added this but I'll check again to confirm that the logic in this part of the algorithm is working correctly.

Reply:
Yeah, my concern was whether the `self.classifier` object will have been fit on the last train/test split from the cross-validation loop for the distribution characteristics calculation, instead of keeping the `self.classifier` object that's originally passed in.